Hide PayPal Standard on new installs

james-allan opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR no longer loads PayPal Standard by default on new sites. This is part of our effort to encourage the adoption of PayPal Payments instead of PayPal Standard. See more @ pb0GrA-1ci-p2

I've used `WC_Install::is_new_install()` so PayPal Standard will be hidden by default for sites that have: 

- No `woocommerce_version`. 
- OR 
   - No products.
   - and have not set up the shop page.

From my tests, that's almost certainly sites that haven't been through the set-up wizard. All other sites will still load PayPal Standard by default.

If a new site wants to bring PayPal Standard back, there is a filter they can use to bypass this condition - `woocommerce_should_load_paypal_standard`.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

### How to test the changes in this Pull Request:

1. On a fresh WP site, install WC using this branch.
2. Either complete the set-up wizard or skip it. 
3. Go to WooCommerce > Settings > Payments. 
4. PayPal Standard won't appear as an option.
5. On an existing store, with this branch. 
3. Go to WooCommerce > Settings > Payments. 
4. PayPal Standard will be listed. 

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

```
* Tweak - No longer load PayPal Standard by default on new installs.
```

james-allan committed:
No longer load PayPal Standard on new installs

james-allan committed:
Set the PayPal load flag on install

james-allan committed:
Load PayPal Standard in unit tests

james-allan commented on includes/class-wc-install.php:
```diff
@@ -310,6 +310,7 @@ public static function install() {
 		self::create_files();
 		self::maybe_create_pages();
 		self::maybe_set_activation_transients();
+		self::set_paypal_standard_load_eligibility();
```
I'm not sure if this is the correct way to go about setting this flag on install. Let me know if there's a preferred way of doing this.

james-allan commented:
The e2e tests are failing because it cannot find PayPal as an option on checkout. I'm not sure how to fix that. For the unit tests, I just attached the `__return_true` callback so it would load, let me know if there's something similar I can do in the e2e tests.

thenbrent commented:
@james-allan a new requirement for this - PayPal Standard should only be hidden when the payment recommendations are displayed.

I believe the banner is displayed only when the marketplace suggestions option is enabled, but haven't confirmed that. 

More context on this requirement: p7bje6-3gv-p2#comment-6294

james-allan commented:
> PayPal Standard should only be hidden when the payment recommendations are displayed.

👍 

So at the moment, PayPal Standard is only hidden for new sites and it's set on install. If that flag isn't set, it will get set anytime WC gets the list of gateways via `WC()->payment_gateways();`. 

So the way I see it there are 2 possible cases:

1. A brand new install. PayPal Standard visibility will be set on install.
2. Existing site. PayPal Standard visibility is set whenever `WC()->payment_gateways()` is called which is pretty frequently and ATM it will always be true (show). 

For new sites, the `woocommerce_show_marketplace_suggestions` option will be 'yes' ([because it's the default](https://github.com/woocommerce/woocommerce/blob/5.3.0/includes/admin/settings/class-wc-settings-advanced.php#L368)) and so checking the visibility of Standard on install will almost certainly result in it being hidden - it's a new install and marketplace suggestions are on (default). Checking the marketplace suggestion option on WC install is just guarenteeing a `true && true`.

As it currently is for existing sites, the `woocommerce_show_marketplace_suggestions` option won't have any bareing on the logic because of the higher order criteria of it not being a new site being the end of the line. 

If we decide to extend this existing sites, then `woocommerce_show_marketplace_suggestions` will be a condition we will need consider. 

I guess my question is. Do I extend this to existing sites, ones with PayPal Stardard not enabled, not set up and with marketplace suggestions on? 

Do I need to recheck the visibility criteria when the `woocommerce_show_marketplace_suggestions` option is changed? For example on a brand new site, PayPal Standard will be hidden, if they then go and disable marketplace suggestions, do I now have to show PayPal Standard?

thenbrent commented:
@james-allan thanks for the clarification.

> Do I extend this to existing sites, ones with PayPal Standard not enabled, not set up and with marketplace suggestions on?

No. For the purposes mentioned in the P2, I think the current approach is even better to hide only on new sites than to hide it only when marektplace suggestions are on. We can start there and decide how to proceed with existing sites in future.

> Do I need to recheck the visibility criteria when the `woocommerce_show_marketplace_suggestions` option is changed? For example on a brand new site PayPal Standard will be hidden, if they then go and disable marketplace suggestions, do I now have to show PayPal Standard?

The only situation that would be problematic in is if the store disables marketplace suggestions before choosing a payment method (either via onboarding wizard or the Settings screen recommendations). I expect that to be rare enough that it doesn't warrant special consideration. So again, the current approach seems good as is.

vedanshujain committed:
Load PayPal gateway directly to prevent potential gateway errors.

`install()` method may be called before wc_version is set, this may cause unknown errors in 3PD gateways.

So its safer to only load PayPal gateway directly.

vedanshujain committed:
No need to check for Paypal standard since its removed.

See #29971 but basically we want to remove PayPal standard in favor of PayPal checkout which has better experience. So this test is not valid anymore.

vedanshujain committed:
Remove PayPal test since its removed by default now.

vedanshujain reviewed (approved):
LGTM! I have added few commits to:
1. Fix failing e2e tests by removing PayPal related checks.
2. Loading only PayPal gateway during setup process. Earlier we were loading "all" the gateways, which may be risky since at the install time not all options would have been set.

zhongruige reviewed (approved):
Looks good! No concerns with the changes to the e2e tests, and they still work great for me locally.

thenbrent commented:
@JimDarker we have not been notified of anything like that, and it seems unlikely given PayPal Standard has been working for over a decade. Do you have a source for where you read that? If that's the case, we'll have work to do on the Woo side.

JimDarker commented:
@thenbrent We were called and emailed by a couple of Inside Sales reps at PayPal saying that we were required to upgrade before March 1st. They stated "The current PayPal solution will not be supported from the 1st of March 2022 by Woo Commerce as per the Woo Commerce website". 

We are definately just on the Standard version but perhaps these employees are mistaken and think we are using PayPal Checkout? Or maybe they're purposefully causing confusion to stimulate the switch over. Either way, things have been a bit unclear and unsettling at our end. If you guys haven't heard anything though that's more encouraging.

thenbrent commented:
@JimDarker thanks for the clarifying. Not being supported is a bit different to it not working.

You will be able to continue to accept payments via PayPal Standard (i.e. it won't stop working), but the PayPal rep is correct that the PayPal Standard and other older PayPal integrations will no longer be maintained or supported (e.g. support tickets via WooCommerce.com, bug fixes etc.).

I recommend upgrading to the newer PayPal integration when possible; however, your store will continue to be able to accept payments via PayPal Standard until you are able to make the switch as far as I know.

user-1138 commented:
@thenbrent [commented](https://github.com/woocommerce/woocommerce/pull/29971#issuecomment-847744962):

> We can start there and decide how to proceed with existing sites in future

I like to clean house, and as I'm migrating existing sites over to the newer/recommended WooCommerce PayPal Payments plugin, I'd also like to hide PayPal Standard from _Settings > Payment methods_, and cleanup WooCoomerce to be like a new/post 5.5.0 site (i.e. one where PayPal Standard was never loaded) or at the very least, remove PayPal Standard related settings/API credentials (which I can obviously at least do manually through the settings page).

And I assume the longer term plan will be to deprecate and then eventually remove PayPal Standard from WooCommerce Core?

Whatever the plan/work for 'hide-pypl-on-existing-sites' is, it  might be worth while establishing an Issue or PR to get it on the record, and update existing [advisories](https://developer.woocommerce.com/2021/07/12/developer-advisory-paypal-standard-will-be-hidden-on-new-installs/) and/or [documentation](https://woocommerce.com/document/woocommerce-paypal-payments/paypal-payments-upgrade-guide/) accordingly so folks know whats happening, thus limiting the flow of further potentially silly comments... which I can only assume will increase as more sites migrate to the newer plugin. :-)

user-1138 commented:
For others looking to hide PayPal Standard on an existing install, here's how I went about it:

In the `wp_options` table, open the option_name `woocommerce_paypal_settings`, and;

If you want to maintain your settings, whilst hiding the gateway, just update the "_should_load" part of the option_value from "yes" to "no" - `s:12:"_should_load";s:2:"no"; `

Or, if you want to hide and revert the gateway options to default WooCommerce 5.5+ values (as though it had never been used), paste the below as the option_value, updating the "email" and "receiver_email" string length/values with your sites own `Settings > General > Administration Email Address` (I believe this is where WooC gets these default values from: `wp_options` table > `admin_email`).

`
a:23:{s:7:"enabled";s:2:"no";s:5:"title";s:6:"PayPal";s:11:"description";s:85:"Pay via PayPal; you can pay with your credit card if you don't have a PayPal account.";s:5:"email";s:15:"ADMIN@EMAIL.COM";s:8:"advanced";s:0:"";s:8:"testmode";s:2:"no";s:5:"debug";s:2:"no";s:16:"ipn_notification";s:3:"yes";s:14:"receiver_email";s:15:"ADMIN@EMAIL.COM";s:14:"identity_token";s:0:"";s:14:"invoice_prefix";s:3:"WC-";s:13:"send_shipping";s:3:"yes";s:16:"address_override";s:2:"no";s:13:"paymentaction";s:4:"sale";s:9:"image_url";s:0:"";s:11:"api_details";s:0:"";s:12:"api_username";s:0:"";s:12:"api_password";s:0:"";s:13:"api_signature";s:0:"";s:20:"sandbox_api_username";s:0:"";s:20:"sandbox_api_password";s:0:"";s:21:"sandbox_api_signature";s:0:"";s:12:"_should_load";s:2:"no";}
`

After that's been updated, jump into `WooCommerce > Settings > Payments` to check PayPal Standard is removed and then click the `Save changes` button at the bottom of the page, which will update the `wp_options`, `woocommerce_gateway_order` value, thus removing the PayPal Standard entry there too.
